packages/test/test-end-to-end-tests/src/test/attachLifecycle.spec.ts

```diff
@@ -0,0 +1,148 @@
+import { requestFluidObject } from "@fluidframework/runtime-utils";
+import {
+    createLoader,
+    ITestFluidObject,
+    timeoutAwait,
+    timeoutPromise,
+} from "@fluidframework/test-utils";
+
+import {generatePairwiseOptions} from "@fluidframework/test-pairwise-generator";
+import { describeFullCompat } from "@fluidframework/test-version-utils";
+import { IResolvedUrl } from "@fluidframework/driver-definitions";
+import { ISharedMap, SharedMap } from "@fluidframework/map";
+import { IFluidHandle } from "@fluidframework/core-interfaces";
+import { AttachState } from "@fluidframework/container-definitions";
+
+//these points are all after creation of all object at which any object can be attached
+const sharedPoints = [3,4,5];
+
+const testConfigs =
+    generatePairwiseOptions({
+        containerAttachPoint:[0, ... sharedPoints],
+        datastoreAttachPoint: [1, ... sharedPoints],
+        datastoreSaveAfterAttach: [true, false],
+        ddsAttachPoint: [2, ... sharedPoints],
+        ddsSaveAfterAttach: [true, false],
+    });
+
+describeFullCompat("Validate Attach lifecycle", (getTestObjectProvider) => {
+    //enumerate test cases, but filter duplicates
+    for(const testConfig of testConfigs.filter((tc)=>
+        tc.containerAttachPoint !== tc.datastoreAttachPoint
+        && tc.containerAttachPoint !== tc.ddsAttachPoint
+        && tc.datastoreAttachPoint !== tc.ddsAttachPoint)
+    ) {
+        it(`Validate attach orders: ${JSON.stringify(testConfig ?? "undefined")}`, async function() {
+            const provider  = getTestObjectProvider();
+            let containerUrl: IResolvedUrl | undefined;
+            {
+                const initLoader = createLoader(
+                    [[provider.defaultCodeDetails, provider.createFluidEntryPoint()]],
+                    provider.documentServiceFactory,
+                    provider.urlResolver,
+                );
+
+                const initContainer = await initLoader.createDetachedContainer(provider.defaultCodeDetails);
+                if(testConfig.containerAttachPoint === 0) {
+                    // point 0 - at container create
+                    await initContainer.attach(provider.driver.createCreateNewRequest(provider.documentId));
+                    containerUrl = initContainer.resolvedUrl;
+                }
+
+                const initDataObject = await requestFluidObject<ITestFluidObject>(initContainer, "default");
+
+                const ds = await initDataObject.context.containerRuntime.createDataStore("default");
+                const newDataObj = await requestFluidObject<ITestFluidObject>(ds, "/");
+                const attachDatastore = async ()=>{
+                    initDataObject.root.set("ds", newDataObj.handle);
+                    while(testConfig.datastoreSaveAfterAttach
+                        && initContainer.isDirty 
+                        && initContainer.attachState !== AttachState.Detached){
```

Comment:
@ssimic2 Just a data point for the other conversation we had, but now i need to consider attach state when evaluating dirty.

Reply:
When the container's not been attached yet, why don't you want to wait for the container to attach and then runtime to save?

Reply:
the container attach can happen after, so we would end up in a deadlock